core/token.go

```diff
@@ -150,3 +154,63 @@ func (ts TokenService) AuthCodeGrant(ctx *HttpContext, userHint string) (ti Toke
 	}
 	return
 }
+
+/* Fetch the public key to validate JWT. Return the key in PEM format or an error if not found. */
+func (ts TokenService) GetPublicKeyPEM(ctx *HttpContext) (pemPublicKey *rsa.PublicKey, err error) {
+	// @todo - we could cache the public key
+	outp := ""
+	if err := ctx.Request("GET", "/SAAS/API/1.0/REST/auth/token?attribute=publicKey&format=pem", nil, &outp); err != nil {
+		return nil, err
+	}
+	pubKey, err := jwt.ParseRSAPublicKeyFromPEM([]byte(outp))
+	return pubKey, err
+}
+
+/* Validate the ID token (locally). Return the list of claims. */
+func (ts TokenService) ValidateIDToken(ctx *HttpContext, idToken string) (map[string]interface{}, error) {
+	if len(idToken) <= 0 {
+		return nil, errors.New("No ID token provided.")
+	}
+
+	// Fetch the public key
+	publicKey, err := ts.GetPublicKeyPEM(ctx)
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Could not fetch public key: %s\n", err.Error()))
+	}
+
+	// Parse takes the token string and a function for looking up the public key
+	token, err := jwt.Parse(idToken, func(token *jwt.Token) (interface{}, error) {
+		ctx.Log.Debug("token claims: %v\n", token.Claims)
+
+		// validate the algorithm we expect
+		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, fmt.Errorf("Unexpected signing method: %v, expect 'RSA256'", token.Header["alg"])
+		}
+		return publicKey, nil
+	})
+
+	if token == nil {
+		return nil, errors.New(fmt.Sprintf("Could not parse the token: %s\n", err))
+	}
+
+	if claims, ok := token.Claims.(jwt.MapClaims); ok && token.Valid {
+		// token is valid (which means claims "exp, iat, nbf" have been validated during the parse),
+		// so now check issuer is valid
+		expectedIssuer := ctx.HostURL + "/SAAS/auth"
+		if !token.Claims.(jwt.MapClaims).VerifyIssuer(expectedIssuer, true) {
+			err = errors.New(fmt.Sprintf("Invalid issuer: '%s', expected '%s", claims["iss"], expectedIssuer))
+		}
+	} else if ve, ok := err.(*jwt.ValidationError); ok {
+		// give more information on why this is not valid
+		if ve.Errors&jwt.ValidationErrorExpired != 0 {
+			fmt.Println("Token is expired")
```

Comment:
In all other commands we only interact with the ctx.Log. This allows us to test err and information output, handle cases for cf plugin, etc. fmt.Println goes to os.Stdout. IMO, moving all user output to this method (as per previous comment) and to use ctx.Log would make this command more readable, testable.